# エラー

## Cognito only valid absolute URLs can be requested エラー

```bash
https://next-auth.js.org/errors#get_authorization_url_error only valid absolute URLs can be requested {
  message: 'only valid absolute URLs can be requested',
  stack: 'TypeError: only valid absolute URLs can be requested\n' +
    '    at Function.request (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/openid-client/lib/helpers/request.js:68:11)\n' +
    '    at Function.discover (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/openid-client/lib/issuer.js:144:38)\n' +
    '    at openidClient (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/lib/oauth/client.js:16:41)\n' +
    '    at getAuthorizationUrl (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/lib/oauth/authorization-url.js:67:51)\n' +
    '    at Object.signin (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/routes/signin.js:37:60)\n' +
    '    at NextAuthHandler (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/index.js:230:39)\n' +
    '    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n' +
    '    at async NextAuthNextHandler (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/next/index.js:23:19)\n' +
    '    at async /Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/next/index.js:59:32\n' +
    '    at async Object.apiResolver (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next/dist/server/api-utils/node.js:179:9)',
  name: 'TypeError'
}
[next-auth][error][SIGNIN_OAUTH_ERROR] 
https://next-auth.js.org/errors#signin_oauth_error only valid absolute URLs can be requested {
  error: {
    message: 'only valid absolute URLs can be requested',
    stack: 'TypeError: only valid absolute URLs can be requested\n' +
      '    at Function.request (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/openid-client/lib/helpers/request.js:68:11)\n' +
      '    at Function.discover (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/openid-client/lib/issuer.js:144:38)\n' +
      '    at openidClient (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/lib/oauth/client.js:16:41)\n' +
      '    at getAuthorizationUrl (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/lib/oauth/authorization-url.js:67:51)\n' +
      '    at Object.signin (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/routes/signin.js:37:60)\n' +
      '    at NextAuthHandler (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/core/index.js:230:39)\n' +
      '    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n' +
      '    at async NextAuthNextHandler (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/next/index.js:23:19)\n' +
      '    at async /Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next-auth/next/index.js:59:32\n' +
      '    at async Object.apiResolver (/Users/sakamoto/Projects/App/next-authorization-app/next-app/node_modules/next/dist/server/api-utils/node.js:179:9)',
    name: 'TypeError'
  },
  provider: {
    id: 'cognito',
    name: 'Cognito',
    type: 'oauth',
    wellKnown: 'next-authorization-sample.auth.ap-northeast-1.amazoncognito.com/.well-known/openid-configuration',
    idToken: true,
    profile: [Function: profile],
    checks: [ 'state' ],
    clientId: '1j7hmg8nnqihut5bt2jcmnkil2',
    clientSecret: '171njsiqkectcvtci5fdp4h4dj13l63i6fl01ejrjpl44b78dag8',
    issuer: 'next-authorization-sample.auth.ap-northeast-1.amazoncognito.com',
    signinUrl: 'http://localhost:3000/api/auth/signin/cognito',
    callbackUrl: 'http://localhost:3000/api/auth/callback/cognito'
  },
  message: 'only valid absolute URLs can be requested'
}
```

> 日本語訳

next-authのみ有効な絶対URLをリクエストできます

> 原因

CognitoProvider設定を間違えていた
使用するライブラリが違ったこと、設定項目が違ったことが原因だった

ClientIDとClientSecretは全般設定 > アプリクライアント > 詳細から見れる
Issuerに関しては`https://cognito-idp.ap-northeast-1.amazonaws.com/[全般設定に記載されているプールID]`を設定する（ap-northeast-1に関しては自分に合ったリージョンに設定する）

```js
// 旧設定
import NextAuth from 'next-auth';
import { Provider } from "next-auth";

export default NextAuth({
    providers: [
        Provider.Cognito({
            cliendId: process.env.COGNITO_CLIENT_ID,
            clientSecret: process.env.COGNITO_CLIENT_SECRET,
            domain: process.env.COGNITO_DOMAIN,
        })
    ]
})

// 新設定
import NextAuth from 'next-auth';
import CognitoProvider from "next-auth/providers/cognito";

export default NextAuth({
  providers: [
    CognitoProvider({
      clientId: process.env.COGNITO_CLIENT_ID,
      clientSecret: process.env.COGNITO_CLIENT_SECRET,
      issuer: process.env.COGNITO_ISSUER,
    })
  ]
})
```





